rhombus/private/map.rkt

```diff
@@ -245,71 +312,130 @@
+                                  (key-comp-map-pair-build-id mapper)
+                                  (key-comp-list->map-id mapper)))))))
+
+(define-for-syntax map-annotation-make-predicate
   (lambda (arg-id predicate-stxs)
     #`(for/and ([(k v) (in-immutable-hash #,arg-id)])
         (and (#,(car predicate-stxs) k)
-             (#,(cadr predicate-stxs) v))))
+             (#,(cadr predicate-stxs) v)))))
+
+(define-for-syntax map-annotation-make-static-info
   (lambda (static-infoss)
     #`((#%index-result #,(cadr static-infoss))
        (#%sequence-element ((#%values (#,(car static-infoss)
-                                       #,(cadr static-infoss)))))))
-  #'map-build-convert #'())
+                                       #,(cadr static-infoss))))))))
+
+(define-annotation-constructor (Map of)
+  ()
+  #'immutable-hash? map-static-infos
+  2
+  #f
+  map-annotation-make-predicate
+  map-annotation-make-static-info
+  #'map-build-convert #'(#hashalw()))
+
+(define-annotation-syntax Map.by
+  (annotation-prefix-operator
+   #'Map.by
+   '((default . stronger))
+   'macro
+   (lambda (stx)
+     (parse-key-comp stx
+                     (lambda (stx arg-stxes str mapper)
+                       (syntax-parse stx
+                         #:datum-literals (op |.| of)
+                         [(form-id (~and dot (op |.|)) (~and of-id of) . tail)
+                          (parse-annotation-of #'(of-id . tail)
+                                               (key-comp-map?-id mapper) map-static-infos
+                                               2 #f
+                                               map-annotation-make-predicate
+                                               map-annotation-make-static-info
+                                               #'map-build-convert #`(#,(key-comp-empty-stx mapper)))]
+                         [(_ . tail)
+                          (values (annotation-predicate-form (key-comp-map?-id mapper)
+                                                             map-static-infos)
+                                  #'tail)]))))))
 
 (define-syntax (map-build-convert arg-id build-convert-stxs kws data)
-  #`(for/fold ([map #hashalw()])
-              ([(k v) (in-immutable-hash #,arg-id)])
-      #:break (not map)
-      (#,(car build-convert-stxs)
-       k
-       (lambda (k)
-         (#,(cadr build-convert-stxs)
-          v
-          (lambda (v)
-            (hash-set map k v))
-          (lambda () #f)))
-       (lambda () #f))))
+  (syntax-parse data
+    [(empty-ht)
+     #`(for/fold ([map empty-ht])
+                 ([(k v) (in-immutable-hash #,arg-id)])
+         #:break (not map)
+         (#,(car build-convert-stxs)
+          k
+          (lambda (k)
+            (#,(cadr build-convert-stxs)
+             v
+             (lambda (v)
+               (hash-set map k v))
+             (lambda () #f)))
+          (lambda () #f)))]))
 
 (define-static-info-syntax Map-pair-build
```

Comment:
The new variants should redirect to this function for static infos, like the set counterparts.